### DISCUSSION TITLE
Swagger OpenAPI json endpoint caching for Ansible dynamic Inventory?  (api/docs/?format=openapi)

### DISCUSSION THREAD

**Discussion by schmitmd** [netbox-community/netbox#5903] [Q&A] [answered]
Hi All, 

I'm seeing some super slow returns on the https://netbox.example.com/api/docs/?format=openapi endpoint.  For me, it's taking almost 15 seconds.  I've got Netbox running behind an NGINX reverse proxy and uWSGI (though I see the same with gunicorn).  Postgres is running on the same server.  I've tried regular proxy_pass with uWSGI doing an HTTP bind and also the native NGINX uwsgi_pass style, with no noticeable difference.  
Maybe someone can help me get how this particular call might be sped up some?  Here's what I've got so far:

I was able to halve the response time (about 6.5s after first GET) by tweaking urls.py, replacing (just testing things out)
`path('api/docs/', schema_view.with_ui('swagger'), name='api_docs'),`
with
`path('api/docs/', schema_view.as_cached_view(cache_timeout=600), name='api_docs'),` and this one had the same effect:
`path('api/docs/', schema_view.without_ui(cache_timeout=600), name='openapi_json'),`

Since the Swagger schema is really only updated between versions of Netbox, is there a way for me to ease the load on the uWSGI process with the endpoint I mentioned?  I don't see much reason to generate it every single time it's GET-requested...

The reason I'm asking is because I'm trying to plug Netbox into Ansible as a dynamic inventory.  This is the location of the Ansible code (which actually forces this specific endpoint's use):
https://github.com/netbox-community/ansible_modules/blob/devel/plugins/inventory/nb_inventory.py#L1114

Can anyone shed some light on how I might get this to perform a little better?

**Comment by ryanmerolle** [comment, marked as answer]
Asible netbox inventory queries a bunch of api endpoints by default.  To speed that up, you can try to trim that out.

For example: if you only want to query physical devices, don't use `query_filters` (since it also queries virtual devices) but use `device_query_filters` 

You could also set `fetch_all` to false and do the same for `services` .

I suggest you read the various arguments on https://docs.ansible.com/ansible/latest/collections/netbox/netbox/nb_inventory_inventory.html#ansible-collections-netbox-netbox-nb-inventory-inventory.

**Comment by schmitmd** [comment]
Isn't that for filtering the results though?  As I understand it, the dynamic inventory first flat-out queries this endpoint: `/api/docs/?format=openapi`.  Then it constructs the filters and makes additional REST calls based on the available resource endpoints and filters it was passed. 
 My indended question here was to ask if we could make the swagger documentation either static or cacheable because it doesn't change between versions of NetBox.  Unless I'm misunderstanding, the swagger doc is being dynamically rendered every single time `/api/docs` is hit.

**Reply by candlerb** [reply]
Yes you are correct.  Downloading the OpenAPI schema always takes place regardless of filter settings, and it's extremely slow because it's generated on demand.

There is an existing feature request for fixing this: see #6423

**Reply by schmitmd** [reply]
lol, looks like one of those "totally independently created the same invention" kind of moments.  We can close out this discussion then & I'll add myself as a watcher to that issue.  thanks!!
